### ISSUE TITLE
[ErrorException] strpos(): Empty needle

### ISSUE THREAD

**Issue by naroga** [icanhazstring/composer-unused#27]
**Describe the bug**
Running `composer unused` on a big project throws an ErrorException.

**Error dump**
```
composer unused -vvv                                                                                                                                                                                       177.74.239.146  1 ↵
Reading ./composer.json
Loading config file /Users/naroga/.composer/config.json
Loading config file /Users/naroga/.composer/auth.json
Loading config file ./composer.json
Checked CA file /private/etc/ssl/cert.pem: valid
Executing command (/Users/naroga/Projects/<redacted>): git branch --no-color --no-abbrev -v
Executing command (/Users/naroga/Projects/<redacted>): git rev-list develop..QU-3465
Reading /Users/naroga/.composer/composer.json
Loading config file /Users/naroga/.composer/config.json
Loading config file /Users/naroga/.composer/auth.json
Loading config file /Users/naroga/.composer/composer.json
Loading config file /Users/naroga/.composer/auth.json
Reading /Users/naroga/.composer/auth.json
Reading /Users/naroga/Projects/<redacted>/vendor/composer/installed.json
Reading /Users/naroga/.composer/vendor/composer/installed.json
Loading plugin Icanhazstring\Composer\Unused\UnusedPlugin
Running 1.7.2 (2018-08-16 16:57:12) with PHP 7.2.10 on Darwin / 18.2.0

Loading packages
----------------

 Loading 57 requirements
 57/57 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100% < 1 sec/< 1 sec 14.0 MiB

 ! [NOTE] Skipped 1 requirements. No package found or invalid constraint.                                               

 * php

 ! [NOTE] Found 56 package(s) to be checked.                                                                            

Scanning files from basedir /Users/naroga/Projects/<redacted>
---------------------------------------------------------------

 5528/5528 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100% 17 secs/17 secs 156.0 MiB


                          
  [ErrorException]        
  strpos(): Empty needle  
                          

Exception trace:
 () at /Users/naroga/.composer/vendor/icanhazstring/composer-unused/src/Subject/PackageSubject.php:29
 Composer\Util\ErrorHandler::handle() at n/a:n/a
 strpos() at /Users/naroga/.composer/vendor/icanhazstring/composer-unused/src/Subject/PackageSubject.php:29
 Icanhazstring\Composer\Unused\Subject\PackageSubject->providesNamespace() at /Users/naroga/.composer/vendor/icanhazstring/composer-unused/src/Command/UnusedCommand.php:121
 Icanhazstring\Composer\Unused\Command\UnusedCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:241
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:843
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:193
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:254
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:103
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:56
 require() at /usr/local/bin/composer:24

unused [-xd|--excludeDir EXCLUDEDIR] [-xp|--excludePackage EXCLUDEPACKAGE]
```

**Additional information**
No apparent memory issues, running on OSX:

```
PHP 7.2.10 (cli) (built: Sep 14 2018 07:07:08) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.10, Copyright (c) 1999-2018, by Zend Technologies
```

**Comment by icanhazstring** [comment]
Hi @naroga thanks for raising an issue.

Two points here:
1. you might have discovered an issue where my error handling doesn't work properly ;)
2. can you send me your `composer.json` if think there might be a psr-0/psr-4 namespace which is an empty string.

Or a required package of yours provides an empty namespace.

**Comment by naroga** [comment]
Hi, @icanhazstring! Sure thing, I'm not sure if I can publicize the contents of our `composer.json` (we have a very strict NDA and even though the contents of `composer.json` alone won't show any business-related information, I don't want to risk legal repercussions).

Is there any way to send you the contents privately?

**Comment by naroga** [comment]
I'm running some debuggers here, give me a couple of minutes.

**Comment by icanhazstring** [comment]
> Hi, @icanhazstring! I'm not sure if I can publicize the contents of our composer.json

No problem there.
As you are looking into the code right now. Try to find a required package, that has no namespace provided in psr-0 oder psr-4 autoload or autoload-dev. Something like this in the `composer.json`

```
"autoload": {
    "psr-4": {
        "": "src/"
    }
}
```

**Comment by naroga** [comment]
Yes, you're right. The packages throwing the exceptions were all private packages, and they're all using empty namespaces.

```
    "autoload-dev": {
        "psr-4": {
            "<redacted>\\Tests\\": "tests/",
            "": "src"
        }
    }
```

**Comment by naroga** [comment]
Adding a verification for empty strings inside the `foreach` on `PackageSubject` fixed it, but I'm not sure if that's the best thing to do.

**Comment by icanhazstring** [comment]
I will probably ignore those packages then and display a message that these packages were ignore and the reason. Related to #24

**Comment by topikito** [comment]
Hi guys - Same issue here.

**Comment by icanhazstring** [comment]
Beside from the obvious problem about the packages with empty namespaces.
The exception handling is a bit wonky. This should be fixed as well.

**Comment by icanhazstring** [comment]
@naroga 

I just tagged a new release `0.3.1`. With this you should be able to run `composer unused -vvv` until the end. You will then hopefully receive a debug dump. Can you confirm this for me?

Next up: Addressing the bigger issue with collecting "failed" packages and display a message to the user.

**Comment by topikito** [comment]
Just writing to confirm with `-vvvv` it works for me now.

**Comment by icanhazstring** [comment]
Solved with `0.4.0`. If the problem persist, please file a new issue.
